src/kit/LogViewer/LogViewer.tsx

```diff
@@ -420,53 +337,39 @@ const LogViewer: React.FC<Props> = ({
     };
   }, [addLogs, decoder, fetchDirection, onError, serverAddress, onFetch, processLogs]);
 
-  // Re-fetch logs when fetch callback changes.
-  useEffect(() => {
-    local.current = clone(defaultLocal);
-
-    setLogs([]);
-    setIsTailing(true);
-    setFetchDirection(FetchDirection.Older);
-  }, [onFetch]);
-
-  // Initialize logs if applicable.
-  useEffect(() => {
-    if (!initialLogs) return;
-
-    addLogs(initialLogs.map((log) => formatLogEntry(decoder(log))));
-  }, [addLogs, decoder, initialLogs]);
-
   // Abort all outstanding API calls if log viewer unmounts.
   useEffect(() => {
     return () => {
       canceler.abort();
     };
   }, [canceler]);
 
-  // Force recomputing messages height when container size changes.
-  useLayoutEffect(() => {
-    if (containerSize.width === 0 || containerSize.height === 0) return;
+  const handleItemsRendered = useCallback(
+    (renderedLogs: ListItem<ViewerLog>[]) => {
+      setShowButtons(renderedLogs.length < logs.length);
+    },
+    [logs.length],
+  );
 
-    const sizeChanged =
-      containerSize.height !== local.current.previousHeight ||
-      containerSize.width !== local.current.previousWidth;
-    if (sizeChanged) resizeLogs();
+  const handleReachedBottom = useCallback(
+    async (atBottom: boolean) => {
+      // May trigger before the initial logs have rendered.
+      if (isFetching || !local.current.isScrollReady) return;
 
-    local.current.previousWidth = containerSize.width;
-    local.current.previousHeight = containerSize.height;
-  }, [containerSize, resizeLogs]);
+      if (atBottom && !isTailing) {
+        const newLogs = await handleFetchMoreLogs('end');
+        if (newLogs?.length === 0) setIsTailing(true);
+      } else if (!atBottom) setIsTailing(false);
+    },
+    [handleFetchMoreLogs, isFetching, isTailing],
+  );
 
-  // Show scrolling buttons based on whether or not logs spill outside of the list view.
-  useLayoutEffect(() => {
-    setShowButtons(() => {
-      if (!logsRef.current || logs.length === 0) return false;
-      const listParent = logsRef.current.firstElementChild;
-      const list = listParent?.firstElementChild;
-      const scrollHeight = list?.scrollHeight ?? 0;
-      const parentHeight = listParent?.clientHeight ?? 0;
-      return scrollHeight > parentHeight;
-    });
-  }, [logs.length, logsRef]);
+  const handleReachedTop = useCallback(
+    (atTop: boolean) => {
+      if (atTop) handleFetchMoreLogs('start');
```

Comment:
when does this trigger if we've prepended 2 million log lines to the virtual list?

Reply:
That function triggers when the user scrolls to the top of the currently loaded logs. There aren't actually 2 million (or billion) items added to the list, that number is only used as an index. My understanding is that internally each item in the list has to have a non-negative index in order to keep track of what should be displayed. For an appending list that's straightforward - just start at 0 and add 1 for each additional item. However, in the case of a prepending list it gets more complicated. In order to ensure that the index never becomes negative as new logs are prepended we offset our starting index by an absurdly large number. Here's the doc page where you can see it in action: https://virtuoso.dev/prepend-items/